File: java/Jenvws/src/main/java/DataObjects/JobData.java
package DataObjects;

import java.io.Serializable;
import java.util.Arrays;
import java.util.UUID;

/**
 * @author jim
 * @class
 * @date 6/27/14
 */
public class JobData implements Serializable {
        /**
         * Gets or sets a Guid for this job.
         */
        private UUID uuid;

        /// <summary>
        /// Gets or sets the name of the .envx project file for this Job.
        /// </summary>
        private String envxName = "UNKNOWN";

        /// <summary>
        /// Gets or sets the uri where the project file(s) can be downloaded from.
        /// </summary>
        private String projectSourceUri = "UNKNOWN";

        /// <summary>
        /// Uri where project results should be uploaded.
        /// </summary>
        private String projectResultsUri = "UNKNOWN";

        /// <summary>
        /// Gets or sets an int array of scenarios that will be run.
        /// </summary>
         
        private int[] projectScenariosArray;

        /// <summary>
        /// Gets or sets a long that is the submit time of the job.
        /// </summary>
         
        private long SubmitTime = Long.MIN_VALUE;

        /// <summary>
        /// Gets or sets a long that is the time this job began executing.
        /// </summary>
         
        public long StartTime = Long.MIN_VALUE;

        /// <summary>
        /// Gets or sets an int array of scenarios that will be run.
        /// </summary>
         
        public long FinishTime = Long.MIN_VALUE;

        /// <summary>
        /// Gets or sets a name submitted by the client to identify the job easier 
        /// (instead of using the guid).
        /// </summary>
         
        public String FriendlyName = "UNKNOWN";

        /// <summary>
        /// Gets or sets the exit code of the envision process.
        /// Initialized to -1.
        /// </summary>
         
        public int EnvisionExitCode = Integer.MIN_VALUE;

        /// <summary>
        /// Gets or sets the tracker's Guid as a String.
        /// </summary>
         
        public UUID TrackerUuid;

        /// <summary>
        /// Gets or sets the private working set size (memory usage) of this job.
        /// </summary>
         
        public long JobPrivateWorkingSetSize = Long.MIN_VALUE;

        public static JobData emptyJob() {
            return _emptyJob;
        }

        private static final JobData _emptyJob = new JobData();

        private JobData() { }

        public JobData(UUID jobId) {
            this.uuid = jobId;
        }


        public static JobData deepCopy(JobData rhs) {

            if (rhs == null) return null;

            JobData copy = new JobData();
            copy.uuid = UUID.fromString(rhs.uuid.toString());
            copy.envxName = rhs.envxName;
            copy.projectSourceUri = rhs.projectSourceUri;
            copy.projectResultsUri = rhs.projectResultsUri;
            copy.projectScenariosArray = Arrays.copyOf(rhs.projectScenariosArray, rhs.projectScenariosArray.length);
            copy.SubmitTime = rhs.SubmitTime;
            copy.FinishTime = rhs.FinishTime;
            copy.FriendlyName = rhs.FriendlyName;
            copy.EnvisionExitCode = rhs.EnvisionExitCode;
            copy.TrackerUuid = UUID.fromString(rhs.uuid.toString());
            copy.JobPrivateWorkingSetSize = rhs.JobPrivateWorkingSetSize;

            return copy;
        }

    public UUID getUuid() {
        return uuid;
    }

    public void setUuid(UUID uuid) {
        this.uuid = uuid;
    }

    public String getEnvxName() {
        return envxName;
    }

    public void setEnvxName(String envxName) {
        this.envxName = envxName;
    }

    public String getProjectSourceUri() {
        return projectSourceUri;
    }

    public void setProjectSourceUri(String projectSourceUri) {
        this.projectSourceUri = projectSourceUri;
    }

    public String getProjectResultsUri() {
        return projectResultsUri;
    }

    public void setProjectResultsUri(String projectResultsUri) {
        this.projectResultsUri = projectResultsUri;
    }

    public int[] getProjectScenariosArray() {
        return projectScenariosArray;
    }

    public void setProjectScenariosArray(int[] projectScenariosArray) {
        this.projectScenariosArray = projectScenariosArray;
    }

    public long getSubmitTime() {
        return SubmitTime;
    }

    public void setSubmitTime(long submitTime) {
        SubmitTime = submitTime;
    }

    public long getStartTime() {
        return StartTime;
    }

    public void setStartTime(long startTime) {
        StartTime = startTime;
    }

    public long getFinishTime() {
        return FinishTime;
    }

    public void setFinishTime(long finishTime) {
        FinishTime = finishTime;
    }

    public String getFriendlyName() {
        return FriendlyName;
    }

    public void setFriendlyName(String friendlyName) {
        FriendlyName = friendlyName;
    }

    public int getEnvisionExitCode() {
        return EnvisionExitCode;
    }

    public void setEnvisionExitCode(int envisionExitCode) {
        EnvisionExitCode = envisionExitCode;
    }

    public UUID getTrackerUuid() {
        return TrackerUuid;
    }

    public void setTrackerUuid(UUID trackerUuid) {
        TrackerUuid = trackerUuid;
    }

    public long getJobPrivateWorkingSetSize() {
        return JobPrivateWorkingSetSize;
    }

    public void setJobPrivateWorkingSetSize(long jobPrivateWorkingSetSize) {
        JobPrivateWorkingSetSize = jobPrivateWorkingSetSize;
    }

    @Override
    public boolean equals(Object o) {

        if (this == o) return true;

        if (o == null || getClass() != o.getClass()) return false;

        JobData jobData = (JobData) o;

        if (!uuid.equals(jobData.uuid)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        return getUuid().hashCode();
    }


}

File: csharp/Envws/EnvwsLib/DataContracts/JobData.cs
using System;
using System.Runtime.Serialization;

namespace EnvwsLib.DataContracts
{
    public enum JobStatus
    {
        QUEUED,           // Queued on orchestrator
        RUNNNG,           // Running on tracker
        COMPLETE,         // Completed

        FAILED           // check the logs (always, the logs!) !
        
    }

    [DataContract]
    public class JobData
    {
        /// <summary>
        /// Gets or sets a Guid for this job.
        /// </summary>
        [DataMember]
        public string Guid { get; set; }

        /// <summary>
        /// Gets or sets the name of the .envx project file for this Job.
        /// </summary>
        [DataMember]
        public string EnvxName { get; set; }

        /// <summary>
        /// Gets or sets the uri where the project file(s) can be downloaded from.
        /// </summary>
        [DataMember]
        public string ProjectSourceUri { get; set; }

        /// <summary>
        /// Uri where project results should be uploaded.
        /// </summary>
//        [DataMember]
//        public string ProjectResultsUri { get; set; }

        /// <summary>
        /// Gets or sets an int array of scenarios that will be run.
        /// </summary>
        [DataMember]
        public int[] ProjectScenarios { get; set; }

        /// <summary>
        /// Gets or sets a long that is the submit time of the job.
        /// </summary>
        [DataMember]
        public long SubmitTime { get; set; }

        /// <summary>
        /// Gets or sets a long that is the time this job began executing.
        /// </summary>
        [DataMember]
        public long StartTime { get; set; }

        /// <summary>
        /// Gets or sets an int array of scenarios that will be run.
        /// </summary>
        [DataMember]
        public long FinishTime { get; set; }

        /// <summary>
        /// Gets or sets a name submitted by the client to identify the job easier 
        /// (instead of using the guid).
        /// </summary>
        [DataMember]
        public string FriendlyName { get; set; }

        /// <summary>
        /// Gets or sets the exit code of the envision process.
        /// Initialized to -1.
        /// </summary>
        [DataMember]
        public int EnvisionExitCode { get; set; }

        /// <summary>
        /// Gets or sets the tracker's Guid as a string.
        /// </summary>
        [DataMember]
        public string TrackerGuid { get; set; }

        [DataMember]
        public JobStatus Status { get; set; }

        /// <summary>
        /// Gets or sets the private working set size (memory usage) of this job.
        /// </summary>
        //TODO: JobPrivateWorkingSetSize not used.
        [DataMember]
        public long JobPrivateWorkingSetSize { get; set; }

        public static JobData EmptyJob { get { return _emptyJob; } }
        private static readonly JobData _emptyJob = new JobData("EMPTY_JOB");

        private JobData(string guid)
        {
            Guid = guid;
            EnvxName = "";
            ProjectSourceUri = "";
            ProjectScenarios = new int[0];
            SubmitTime = 0L;
            FinishTime = 0L;
            StartTime = 0L;
            FriendlyName = "";
            EnvisionExitCode = -1;
            TrackerGuid = "";
            JobPrivateWorkingSetSize = -1L;
        }

        public JobData() : this("") { }


        public static JobData DeepCopy(JobData rhs)
        {
            if (rhs == null) return null;

            JobData copy = new JobData()
            {
                Guid = string.Copy(rhs.Guid),
                EnvxName = string.Copy(rhs.EnvxName),
                ProjectSourceUri = string.Copy(rhs.ProjectSourceUri),
                SubmitTime = rhs.SubmitTime,
                FinishTime = rhs.FinishTime,
                StartTime = rhs.StartTime,
                FriendlyName = string.Copy(rhs.FriendlyName),
                EnvisionExitCode = rhs.EnvisionExitCode,
                TrackerGuid = string.Copy(rhs.TrackerGuid),
                JobPrivateWorkingSetSize = rhs.JobPrivateWorkingSetSize
            };

            copy.ProjectScenarios = new int[rhs.ProjectScenarios.Length];
            Array.Copy(rhs.ProjectScenarios, copy.ProjectScenarios, copy.ProjectScenarios.Length);
            
            return copy;
        }

        public override bool Equals(object obj)
        {
            if (obj == null || GetType() != obj.GetType())
            {
                return false;
            }

            JobData j = obj as JobData;
            return Guid.Equals(j.Guid);
        }

        public override int GetHashCode()
        {
            return base.GetHashCode();
        }
    }
}
